// @pulumi/aws/apigateway/resource.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an API Gateway Resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const myDemoAPI = new aws.apigateway.RestApi("myDemoAPI", {description: "This is my API for demonstration purposes"});
 * const myDemoResource = new aws.apigateway.Resource("myDemoResource", {
 *     restApi: myDemoAPI.id,
 *     parentId: myDemoAPI.rootResourceId,
 *     pathPart: "mydemoresource",
 * });
 * ```
 *
 * ## Import
 *
 * `aws_api_gateway_resource` can be imported using `REST-API-ID/RESOURCE-ID`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:apigateway/resource:Resource example 12345abcde/67890fghij
 * ```
 */
class Resource extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["parentId"] = state ? state.parentId : undefined;
            inputs["path"] = state ? state.path : undefined;
            inputs["pathPart"] = state ? state.pathPart : undefined;
            inputs["restApi"] = state ? state.restApi : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.parentId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'parentId'");
            }
            if ((!args || args.pathPart === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'pathPart'");
            }
            if ((!args || args.restApi === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'restApi'");
            }
            inputs["parentId"] = args ? args.parentId : undefined;
            inputs["pathPart"] = args ? args.pathPart : undefined;
            inputs["restApi"] = args ? args.restApi : undefined;
            inputs["path"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Resource.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Resource resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Resource(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Resource.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Resource.__pulumiType;
    }
}
exports.Resource = Resource;
/** @internal */
Resource.__pulumiType = 'aws:apigateway/resource:Resource';
