// @pulumi/aws/datasync/locationFsxWindows.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an AWS DataSync FSx Windows Location.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.datasync.LocationFsxWindows("example", {
 *     fsxFilesystemArn: aws_fsx_windows_file_system.example.arn,
 *     user: "SomeUser",
 *     password: "SuperSecretPassw0rd",
 *     securityGroupArns: [aws_security_group.example.arn],
 * });
 * ```
 *
 * ## Import
 *
 * `aws_datasync_location_fsx_windows_file_system` can be imported by using the `DataSync-ARN#FSx-Windows-ARN`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:datasync/locationFsxWindows:LocationFsxWindows example arn:aws:datasync:us-west-2:123456789012:location/loc-12345678901234567#arn:aws:fsx:us-west-2:476956259333:file-system/fs-08e04cd442c1bb94a
 * ```
 */
class LocationFsxWindows extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["creationTime"] = state ? state.creationTime : undefined;
            inputs["domain"] = state ? state.domain : undefined;
            inputs["fsxFilesystemArn"] = state ? state.fsxFilesystemArn : undefined;
            inputs["password"] = state ? state.password : undefined;
            inputs["securityGroupArns"] = state ? state.securityGroupArns : undefined;
            inputs["subdirectory"] = state ? state.subdirectory : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["uri"] = state ? state.uri : undefined;
            inputs["user"] = state ? state.user : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.fsxFilesystemArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'fsxFilesystemArn'");
            }
            if ((!args || args.password === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'password'");
            }
            if ((!args || args.securityGroupArns === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'securityGroupArns'");
            }
            if ((!args || args.user === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'user'");
            }
            inputs["domain"] = args ? args.domain : undefined;
            inputs["fsxFilesystemArn"] = args ? args.fsxFilesystemArn : undefined;
            inputs["password"] = args ? args.password : undefined;
            inputs["securityGroupArns"] = args ? args.securityGroupArns : undefined;
            inputs["subdirectory"] = args ? args.subdirectory : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["user"] = args ? args.user : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["creationTime"] = undefined /*out*/;
            inputs["uri"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(LocationFsxWindows.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing LocationFsxWindows resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new LocationFsxWindows(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of LocationFsxWindows.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === LocationFsxWindows.__pulumiType;
    }
}
exports.LocationFsxWindows = LocationFsxWindows;
/** @internal */
LocationFsxWindows.__pulumiType = 'aws:datasync/locationFsxWindows:LocationFsxWindows';
