// @pulumi/aws/xray/samplingRule.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Creates and manages an AWS XRay Sampling Rule.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.xray.SamplingRule("example", {
 *     attributes: {
 *         Hello: "Tris",
 *     },
 *     fixedRate: 0.05,
 *     host: "*",
 *     httpMethod: "*",
 *     priority: 10000,
 *     reservoirSize: 1,
 *     resourceArn: "*",
 *     ruleName: "example",
 *     serviceName: "*",
 *     serviceType: "*",
 *     urlPath: "*",
 *     version: 1,
 * });
 * ```
 *
 * ## Import
 *
 * XRay Sampling Rules can be imported using the name, e.g.
 *
 * ```sh
 *  $ pulumi import aws:xray/samplingRule:SamplingRule example example
 * ```
 */
class SamplingRule extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["attributes"] = state ? state.attributes : undefined;
            inputs["fixedRate"] = state ? state.fixedRate : undefined;
            inputs["host"] = state ? state.host : undefined;
            inputs["httpMethod"] = state ? state.httpMethod : undefined;
            inputs["priority"] = state ? state.priority : undefined;
            inputs["reservoirSize"] = state ? state.reservoirSize : undefined;
            inputs["resourceArn"] = state ? state.resourceArn : undefined;
            inputs["ruleName"] = state ? state.ruleName : undefined;
            inputs["serviceName"] = state ? state.serviceName : undefined;
            inputs["serviceType"] = state ? state.serviceType : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["urlPath"] = state ? state.urlPath : undefined;
            inputs["version"] = state ? state.version : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.fixedRate === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'fixedRate'");
            }
            if ((!args || args.host === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'host'");
            }
            if ((!args || args.httpMethod === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'httpMethod'");
            }
            if ((!args || args.priority === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'priority'");
            }
            if ((!args || args.reservoirSize === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'reservoirSize'");
            }
            if ((!args || args.resourceArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'resourceArn'");
            }
            if ((!args || args.serviceName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'serviceName'");
            }
            if ((!args || args.serviceType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'serviceType'");
            }
            if ((!args || args.urlPath === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'urlPath'");
            }
            if ((!args || args.version === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'version'");
            }
            inputs["attributes"] = args ? args.attributes : undefined;
            inputs["fixedRate"] = args ? args.fixedRate : undefined;
            inputs["host"] = args ? args.host : undefined;
            inputs["httpMethod"] = args ? args.httpMethod : undefined;
            inputs["priority"] = args ? args.priority : undefined;
            inputs["reservoirSize"] = args ? args.reservoirSize : undefined;
            inputs["resourceArn"] = args ? args.resourceArn : undefined;
            inputs["ruleName"] = args ? args.ruleName : undefined;
            inputs["serviceName"] = args ? args.serviceName : undefined;
            inputs["serviceType"] = args ? args.serviceType : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["urlPath"] = args ? args.urlPath : undefined;
            inputs["version"] = args ? args.version : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(SamplingRule.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing SamplingRule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new SamplingRule(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of SamplingRule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SamplingRule.__pulumiType;
    }
}
exports.SamplingRule = SamplingRule;
/** @internal */
SamplingRule.__pulumiType = 'aws:xray/samplingRule:SamplingRule';
