// @pulumi/aws/cloudwatch/eventPermission.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a resource to create an EventBridge permission to support cross-account events in the current account default event bus.
 *
 * > **Note:** EventBridge was formerly known as CloudWatch Events. The functionality is identical.
 *
 * ## Example Usage
 * ### Account Access
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const devAccountAccess = new aws.cloudwatch.EventPermission("DevAccountAccess", {
 *     principal: "123456789012",
 *     statementId: "DevAccountAccess",
 * });
 * ```
 * ### Organization Access
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const organizationAccess = new aws.cloudwatch.EventPermission("organizationAccess", {
 *     principal: "*",
 *     statementId: "OrganizationAccess",
 *     condition: {
 *         key: "aws:PrincipalOrgID",
 *         type: "StringEquals",
 *         value: aws_organizations_organization.example.id,
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * EventBridge permissions can be imported using the `event_bus_name/statement_id` (if you omit `event_bus_name`, the `default` event bus will be used), e.g.
 *
 * ```sh
 *  $ pulumi import aws:cloudwatch/eventPermission:EventPermission DevAccountAccess example-event-bus/DevAccountAccess
 * ```
 */
class EventPermission extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["action"] = state ? state.action : undefined;
            inputs["condition"] = state ? state.condition : undefined;
            inputs["eventBusName"] = state ? state.eventBusName : undefined;
            inputs["principal"] = state ? state.principal : undefined;
            inputs["statementId"] = state ? state.statementId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.principal === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'principal'");
            }
            if ((!args || args.statementId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'statementId'");
            }
            inputs["action"] = args ? args.action : undefined;
            inputs["condition"] = args ? args.condition : undefined;
            inputs["eventBusName"] = args ? args.eventBusName : undefined;
            inputs["principal"] = args ? args.principal : undefined;
            inputs["statementId"] = args ? args.statementId : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(EventPermission.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing EventPermission resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new EventPermission(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of EventPermission.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EventPermission.__pulumiType;
    }
}
exports.EventPermission = EventPermission;
/** @internal */
EventPermission.__pulumiType = 'aws:cloudwatch/eventPermission:EventPermission';
