// @pulumi/aws/quicksight/user.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing QuickSight User
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.quicksight.User("example", {
 *     email: "author@example.com",
 *     identityType: "IAM",
 *     userName: "an-author",
 *     userRole: "AUTHOR",
 * });
 * ```
 *
 * ## Import
 *
 * Importing is currently not supported on this resource.
 */
class User extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["awsAccountId"] = state ? state.awsAccountId : undefined;
            inputs["email"] = state ? state.email : undefined;
            inputs["iamArn"] = state ? state.iamArn : undefined;
            inputs["identityType"] = state ? state.identityType : undefined;
            inputs["namespace"] = state ? state.namespace : undefined;
            inputs["sessionName"] = state ? state.sessionName : undefined;
            inputs["userName"] = state ? state.userName : undefined;
            inputs["userRole"] = state ? state.userRole : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.email === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'email'");
            }
            if ((!args || args.identityType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'identityType'");
            }
            if ((!args || args.userRole === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'userRole'");
            }
            inputs["awsAccountId"] = args ? args.awsAccountId : undefined;
            inputs["email"] = args ? args.email : undefined;
            inputs["iamArn"] = args ? args.iamArn : undefined;
            inputs["identityType"] = args ? args.identityType : undefined;
            inputs["namespace"] = args ? args.namespace : undefined;
            inputs["sessionName"] = args ? args.sessionName : undefined;
            inputs["userName"] = args ? args.userName : undefined;
            inputs["userRole"] = args ? args.userRole : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(User.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing User resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new User(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of User.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === User.__pulumiType;
    }
}
exports.User = User;
/** @internal */
User.__pulumiType = 'aws:quicksight/user:User';
