// @pulumi/aws/ses/identityNotificationTopic.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing SES Identity Notification Topics
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.ses.IdentityNotificationTopic("test", {
 *     topicArn: aws_sns_topic.example.arn,
 *     notificationType: "Bounce",
 *     identity: aws_ses_domain_identity.example.domain,
 *     includeOriginalHeaders: true,
 * });
 * ```
 *
 * ## Import
 *
 * Identity Notification Topics can be imported using ID of the record. The ID is made up as IDENTITY|TYPE where IDENTITY is the SES Identity and TYPE is the Notification Type. e.g. example.com|Bounce In this example, `example.com` is the SES Identity and `Bounce` is the Notification Type. To import the ID above, it would look as follows
 *
 * ```sh
 *  $ pulumi import aws:ses/identityNotificationTopic:IdentityNotificationTopic test 'example.com|Bounce'
 * ```
 */
class IdentityNotificationTopic extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["identity"] = state ? state.identity : undefined;
            inputs["includeOriginalHeaders"] = state ? state.includeOriginalHeaders : undefined;
            inputs["notificationType"] = state ? state.notificationType : undefined;
            inputs["topicArn"] = state ? state.topicArn : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.identity === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'identity'");
            }
            if ((!args || args.notificationType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'notificationType'");
            }
            inputs["identity"] = args ? args.identity : undefined;
            inputs["includeOriginalHeaders"] = args ? args.includeOriginalHeaders : undefined;
            inputs["notificationType"] = args ? args.notificationType : undefined;
            inputs["topicArn"] = args ? args.topicArn : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(IdentityNotificationTopic.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing IdentityNotificationTopic resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new IdentityNotificationTopic(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of IdentityNotificationTopic.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === IdentityNotificationTopic.__pulumiType;
    }
}
exports.IdentityNotificationTopic = IdentityNotificationTopic;
/** @internal */
IdentityNotificationTopic.__pulumiType = 'aws:ses/identityNotificationTopic:IdentityNotificationTopic';
