// @pulumi/aws/workspaces/workspace.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a workspace in [AWS Workspaces](https://docs.aws.amazon.com/workspaces/latest/adminguide/amazon-workspaces.html) Service
 *
 * > **NOTE:** AWS WorkSpaces service requires [`workspaces_DefaultRole`](https://docs.aws.amazon.com/workspaces/latest/adminguide/workspaces-access-control.html#create-default-role) IAM role to operate normally.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const valueWindows10 = aws.workspaces.getBundle({
 *     bundleId: "wsb-bh8rsxt14",
 * });
 * const example = new aws.workspaces.Workspace("example", {
 *     directoryId: aws_workspaces_directory.example.id,
 *     bundleId: valueWindows10.then(valueWindows10 => valueWindows10.id),
 *     userName: "john.doe",
 *     rootVolumeEncryptionEnabled: true,
 *     userVolumeEncryptionEnabled: true,
 *     volumeEncryptionKey: "alias/aws/workspaces",
 *     workspaceProperties: {
 *         computeTypeName: "VALUE",
 *         userVolumeSizeGib: 10,
 *         rootVolumeSizeGib: 80,
 *         runningMode: "AUTO_STOP",
 *         runningModeAutoStopTimeoutInMinutes: 60,
 *     },
 *     tags: {
 *         Department: "IT",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Workspaces can be imported using their ID, e.g.
 *
 * ```sh
 *  $ pulumi import aws:workspaces/workspace:Workspace example ws-9z9zmbkhv
 * ```
 */
class Workspace extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["bundleId"] = state ? state.bundleId : undefined;
            inputs["computerName"] = state ? state.computerName : undefined;
            inputs["directoryId"] = state ? state.directoryId : undefined;
            inputs["ipAddress"] = state ? state.ipAddress : undefined;
            inputs["rootVolumeEncryptionEnabled"] = state ? state.rootVolumeEncryptionEnabled : undefined;
            inputs["state"] = state ? state.state : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["userName"] = state ? state.userName : undefined;
            inputs["userVolumeEncryptionEnabled"] = state ? state.userVolumeEncryptionEnabled : undefined;
            inputs["volumeEncryptionKey"] = state ? state.volumeEncryptionKey : undefined;
            inputs["workspaceProperties"] = state ? state.workspaceProperties : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.bundleId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'bundleId'");
            }
            if ((!args || args.directoryId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'directoryId'");
            }
            if ((!args || args.userName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'userName'");
            }
            inputs["bundleId"] = args ? args.bundleId : undefined;
            inputs["directoryId"] = args ? args.directoryId : undefined;
            inputs["rootVolumeEncryptionEnabled"] = args ? args.rootVolumeEncryptionEnabled : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["userName"] = args ? args.userName : undefined;
            inputs["userVolumeEncryptionEnabled"] = args ? args.userVolumeEncryptionEnabled : undefined;
            inputs["volumeEncryptionKey"] = args ? args.volumeEncryptionKey : undefined;
            inputs["workspaceProperties"] = args ? args.workspaceProperties : undefined;
            inputs["computerName"] = undefined /*out*/;
            inputs["ipAddress"] = undefined /*out*/;
            inputs["state"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Workspace.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Workspace resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Workspace(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Workspace.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Workspace.__pulumiType;
    }
}
exports.Workspace = Workspace;
/** @internal */
Workspace.__pulumiType = 'aws:workspaces/workspace:Workspace';
