// @pulumi/aws/ecr/repository.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an Elastic Container Registry Repository.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const foo = new aws.ecr.Repository("foo", {
 *     imageScanningConfiguration: {
 *         scanOnPush: true,
 *     },
 *     imageTagMutability: "MUTABLE",
 * });
 * ```
 *
 * ## Import
 *
 * ECR Repositories can be imported using the `name`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:ecr/repository:Repository service test-service
 * ```
 */
class Repository extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["encryptionConfigurations"] = state ? state.encryptionConfigurations : undefined;
            inputs["imageScanningConfiguration"] = state ? state.imageScanningConfiguration : undefined;
            inputs["imageTagMutability"] = state ? state.imageTagMutability : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["registryId"] = state ? state.registryId : undefined;
            inputs["repositoryUrl"] = state ? state.repositoryUrl : undefined;
            inputs["tags"] = state ? state.tags : undefined;
        }
        else {
            const args = argsOrState;
            inputs["encryptionConfigurations"] = args ? args.encryptionConfigurations : undefined;
            inputs["imageScanningConfiguration"] = args ? args.imageScanningConfiguration : undefined;
            inputs["imageTagMutability"] = args ? args.imageTagMutability : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["registryId"] = undefined /*out*/;
            inputs["repositoryUrl"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Repository.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Repository resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Repository(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Repository.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Repository.__pulumiType;
    }
}
exports.Repository = Repository;
/** @internal */
Repository.__pulumiType = 'aws:ecr/repository:Repository';
