// @pulumi/aws/codeartifact/domainPermissions.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a CodeArtifact Domains Permissions Policy Resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const exampleKey = new aws.kms.Key("exampleKey", {description: "domain key"});
 * const exampleDomain = new aws.codeartifact.Domain("exampleDomain", {
 *     domain: "example.com",
 *     encryptionKey: exampleKey.arn,
 * });
 * const test = new aws.codeartifact.DomainPermissions("test", {
 *     domain: exampleDomain.domain,
 *     policyDocument: pulumi.interpolate`{
 *     "Version": "2012-10-17",
 *     "Statement": [
 *         {
 *             "Action": "codeartifact:CreateRepository",
 *             "Effect": "Allow",
 *             "Principal": "*",
 *             "Resource": "${exampleDomain.arn}"
 *         }
 *     ]
 * }
 * `,
 * });
 * ```
 *
 * ## Import
 *
 * CodeArtifact Domain Permissions Policies can be imported using the CodeArtifact Domain ARN, e.g.
 *
 * ```sh
 *  $ pulumi import aws:codeartifact/domainPermissions:DomainPermissions example arn:aws:codeartifact:us-west-2:012345678912:domain/tf-acc-test-1928056699409417367
 * ```
 */
class DomainPermissions extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["domain"] = state ? state.domain : undefined;
            inputs["domainOwner"] = state ? state.domainOwner : undefined;
            inputs["policyDocument"] = state ? state.policyDocument : undefined;
            inputs["policyRevision"] = state ? state.policyRevision : undefined;
            inputs["resourceArn"] = state ? state.resourceArn : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.domain === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'domain'");
            }
            if ((!args || args.policyDocument === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'policyDocument'");
            }
            inputs["domain"] = args ? args.domain : undefined;
            inputs["domainOwner"] = args ? args.domainOwner : undefined;
            inputs["policyDocument"] = args ? args.policyDocument : undefined;
            inputs["policyRevision"] = args ? args.policyRevision : undefined;
            inputs["resourceArn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(DomainPermissions.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing DomainPermissions resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new DomainPermissions(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of DomainPermissions.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === DomainPermissions.__pulumiType;
    }
}
exports.DomainPermissions = DomainPermissions;
/** @internal */
DomainPermissions.__pulumiType = 'aws:codeartifact/domainPermissions:DomainPermissions';
