// @pulumi/aws/codeartifact/domain.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a CodeArtifact Domain Resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const exampleKey = new aws.kms.Key("exampleKey", {description: "domain key"});
 * const exampleDomain = new aws.codeartifact.Domain("exampleDomain", {
 *     domain: "example",
 *     encryptionKey: exampleKey.arn,
 * });
 * ```
 *
 * ## Import
 *
 * CodeArtifact Domain can be imported using the CodeArtifact Domain arn, e.g.
 *
 * ```sh
 *  $ pulumi import aws:codeartifact/domain:Domain example arn:aws:codeartifact:us-west-2:012345678912:domain/tf-acc-test-8593714120730241305
 * ```
 */
class Domain extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["assetSizeBytes"] = state ? state.assetSizeBytes : undefined;
            inputs["createdTime"] = state ? state.createdTime : undefined;
            inputs["domain"] = state ? state.domain : undefined;
            inputs["encryptionKey"] = state ? state.encryptionKey : undefined;
            inputs["owner"] = state ? state.owner : undefined;
            inputs["repositoryCount"] = state ? state.repositoryCount : undefined;
            inputs["tags"] = state ? state.tags : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.domain === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'domain'");
            }
            if ((!args || args.encryptionKey === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'encryptionKey'");
            }
            inputs["domain"] = args ? args.domain : undefined;
            inputs["encryptionKey"] = args ? args.encryptionKey : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["assetSizeBytes"] = undefined /*out*/;
            inputs["createdTime"] = undefined /*out*/;
            inputs["owner"] = undefined /*out*/;
            inputs["repositoryCount"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Domain.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Domain resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Domain(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Domain.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Domain.__pulumiType;
    }
}
exports.Domain = Domain;
/** @internal */
Domain.__pulumiType = 'aws:codeartifact/domain:Domain';
