// @pulumi/aws/opsworks/permission.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an OpsWorks permission resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const myStackPermission = new aws.opsworks.Permission("myStackPermission", {
 *     allowSsh: true,
 *     allowSudo: true,
 *     level: "iam_only",
 *     userArn: aws_iam_user.user.arn,
 *     stackId: aws_opsworks_stack.stack.id,
 * });
 * ```
 */
class Permission extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["allowSsh"] = state ? state.allowSsh : undefined;
            inputs["allowSudo"] = state ? state.allowSudo : undefined;
            inputs["level"] = state ? state.level : undefined;
            inputs["stackId"] = state ? state.stackId : undefined;
            inputs["userArn"] = state ? state.userArn : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.userArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'userArn'");
            }
            inputs["allowSsh"] = args ? args.allowSsh : undefined;
            inputs["allowSudo"] = args ? args.allowSudo : undefined;
            inputs["level"] = args ? args.level : undefined;
            inputs["stackId"] = args ? args.stackId : undefined;
            inputs["userArn"] = args ? args.userArn : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Permission.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Permission resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Permission(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Permission.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Permission.__pulumiType;
    }
}
exports.Permission = Permission;
/** @internal */
Permission.__pulumiType = 'aws:opsworks/permission:Permission';
