// @pulumi/aws/opsworks/stack.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an OpsWorks stack resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const main = new aws.opsworks.Stack("main", {
 *     region: "us-west-1",
 *     serviceRoleArn: aws_iam_role.opsworks.arn,
 *     defaultInstanceProfileArn: aws_iam_instance_profile.opsworks.arn,
 *     tags: {
 *         Name: "foobar-stack",
 *     },
 *     customJson: `{
 *  "foobar": {
 *     "version": "1.0.0"
 *   }
 * }
 * `,
 * });
 * ```
 *
 * ## Import
 *
 * OpsWorks stacks can be imported using the `id`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:opsworks/stack:Stack bar 00000000-0000-0000-0000-000000000000
 * ```
 */
class Stack extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["agentVersion"] = state ? state.agentVersion : undefined;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["berkshelfVersion"] = state ? state.berkshelfVersion : undefined;
            inputs["color"] = state ? state.color : undefined;
            inputs["configurationManagerName"] = state ? state.configurationManagerName : undefined;
            inputs["configurationManagerVersion"] = state ? state.configurationManagerVersion : undefined;
            inputs["customCookbooksSources"] = state ? state.customCookbooksSources : undefined;
            inputs["customJson"] = state ? state.customJson : undefined;
            inputs["defaultAvailabilityZone"] = state ? state.defaultAvailabilityZone : undefined;
            inputs["defaultInstanceProfileArn"] = state ? state.defaultInstanceProfileArn : undefined;
            inputs["defaultOs"] = state ? state.defaultOs : undefined;
            inputs["defaultRootDeviceType"] = state ? state.defaultRootDeviceType : undefined;
            inputs["defaultSshKeyName"] = state ? state.defaultSshKeyName : undefined;
            inputs["defaultSubnetId"] = state ? state.defaultSubnetId : undefined;
            inputs["hostnameTheme"] = state ? state.hostnameTheme : undefined;
            inputs["manageBerkshelf"] = state ? state.manageBerkshelf : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["region"] = state ? state.region : undefined;
            inputs["serviceRoleArn"] = state ? state.serviceRoleArn : undefined;
            inputs["stackEndpoint"] = state ? state.stackEndpoint : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["useCustomCookbooks"] = state ? state.useCustomCookbooks : undefined;
            inputs["useOpsworksSecurityGroups"] = state ? state.useOpsworksSecurityGroups : undefined;
            inputs["vpcId"] = state ? state.vpcId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.defaultInstanceProfileArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'defaultInstanceProfileArn'");
            }
            if ((!args || args.region === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'region'");
            }
            if ((!args || args.serviceRoleArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'serviceRoleArn'");
            }
            inputs["agentVersion"] = args ? args.agentVersion : undefined;
            inputs["berkshelfVersion"] = args ? args.berkshelfVersion : undefined;
            inputs["color"] = args ? args.color : undefined;
            inputs["configurationManagerName"] = args ? args.configurationManagerName : undefined;
            inputs["configurationManagerVersion"] = args ? args.configurationManagerVersion : undefined;
            inputs["customCookbooksSources"] = args ? args.customCookbooksSources : undefined;
            inputs["customJson"] = args ? args.customJson : undefined;
            inputs["defaultAvailabilityZone"] = args ? args.defaultAvailabilityZone : undefined;
            inputs["defaultInstanceProfileArn"] = args ? args.defaultInstanceProfileArn : undefined;
            inputs["defaultOs"] = args ? args.defaultOs : undefined;
            inputs["defaultRootDeviceType"] = args ? args.defaultRootDeviceType : undefined;
            inputs["defaultSshKeyName"] = args ? args.defaultSshKeyName : undefined;
            inputs["defaultSubnetId"] = args ? args.defaultSubnetId : undefined;
            inputs["hostnameTheme"] = args ? args.hostnameTheme : undefined;
            inputs["manageBerkshelf"] = args ? args.manageBerkshelf : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["region"] = args ? args.region : undefined;
            inputs["serviceRoleArn"] = args ? args.serviceRoleArn : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["useCustomCookbooks"] = args ? args.useCustomCookbooks : undefined;
            inputs["useOpsworksSecurityGroups"] = args ? args.useOpsworksSecurityGroups : undefined;
            inputs["vpcId"] = args ? args.vpcId : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["stackEndpoint"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Stack.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Stack resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Stack(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Stack.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Stack.__pulumiType;
    }
}
exports.Stack = Stack;
/** @internal */
Stack.__pulumiType = 'aws:opsworks/stack:Stack';
