// @pulumi/aws/sagemaker/model.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a SageMaker model resource.
 *
 * ## Example Usage
 *
 * Basic usage:
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const assumeRole = aws.iam.getPolicyDocument({
 *     statements: [{
 *         actions: ["sts:AssumeRole"],
 *         principals: [{
 *             type: "Service",
 *             identifiers: ["sagemaker.amazonaws.com"],
 *         }],
 *     }],
 * });
 * const exampleRole = new aws.iam.Role("exampleRole", {assumeRolePolicy: assumeRole.then(assumeRole => assumeRole.json)});
 * const exampleModel = new aws.sagemaker.Model("exampleModel", {
 *     executionRoleArn: exampleRole.arn,
 *     primaryContainer: {
 *         image: "174872318107.dkr.ecr.us-west-2.amazonaws.com/kmeans:1",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Models can be imported using the `name`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:sagemaker/model:Model test_model model-foo
 * ```
 */
class Model extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["containers"] = state ? state.containers : undefined;
            inputs["enableNetworkIsolation"] = state ? state.enableNetworkIsolation : undefined;
            inputs["executionRoleArn"] = state ? state.executionRoleArn : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["primaryContainer"] = state ? state.primaryContainer : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["vpcConfig"] = state ? state.vpcConfig : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.executionRoleArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'executionRoleArn'");
            }
            inputs["containers"] = args ? args.containers : undefined;
            inputs["enableNetworkIsolation"] = args ? args.enableNetworkIsolation : undefined;
            inputs["executionRoleArn"] = args ? args.executionRoleArn : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["primaryContainer"] = args ? args.primaryContainer : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["vpcConfig"] = args ? args.vpcConfig : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Model.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Model resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Model(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Model.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Model.__pulumiType;
    }
}
exports.Model = Model;
/** @internal */
Model.__pulumiType = 'aws:sagemaker/model:Model';
