// @pulumi/docker/service.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * ## Import
 *
 * Docker service can be imported using the long id, e.g. for a service with the short id `55ba873dd`
 *
 * ```sh
 *  $ pulumi import docker:index/service:Service foo $(docker service inspect -f {{.ID}} 55b)
 * ```
 */
class Service extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["auth"] = state ? state.auth : undefined;
            inputs["convergeConfig"] = state ? state.convergeConfig : undefined;
            inputs["endpointSpec"] = state ? state.endpointSpec : undefined;
            inputs["labels"] = state ? state.labels : undefined;
            inputs["mode"] = state ? state.mode : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["rollbackConfig"] = state ? state.rollbackConfig : undefined;
            inputs["taskSpec"] = state ? state.taskSpec : undefined;
            inputs["updateConfig"] = state ? state.updateConfig : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.taskSpec === undefined) {
                throw new Error("Missing required property 'taskSpec'");
            }
            inputs["auth"] = args ? args.auth : undefined;
            inputs["convergeConfig"] = args ? args.convergeConfig : undefined;
            inputs["endpointSpec"] = args ? args.endpointSpec : undefined;
            inputs["labels"] = args ? args.labels : undefined;
            inputs["mode"] = args ? args.mode : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["rollbackConfig"] = args ? args.rollbackConfig : undefined;
            inputs["taskSpec"] = args ? args.taskSpec : undefined;
            inputs["updateConfig"] = args ? args.updateConfig : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Service.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Service resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Service(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Service.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Service.__pulumiType;
    }
}
exports.Service = Service;
/** @internal */
Service.__pulumiType = 'docker:index/service:Service';
