// @pulumi/aws/cognito/identityProvider.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Cognito User Identity Provider resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.cognito.UserPool("example", {autoVerifiedAttributes: ["email"]});
 * const exampleProvider = new aws.cognito.IdentityProvider("exampleProvider", {
 *     userPoolId: example.id,
 *     providerName: "Google",
 *     providerType: "Google",
 *     providerDetails: {
 *         authorize_scopes: "email",
 *         client_id: "your client_id",
 *         client_secret: "your client_secret",
 *     },
 *     attributeMapping: {
 *         email: "email",
 *         username: "sub",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * `aws_cognito_identity_provider` resources can be imported using their User Pool ID and Provider Name, e.g.
 *
 * ```sh
 *  $ pulumi import aws:cognito/identityProvider:IdentityProvider example xxx_yyyyy:example
 * ```
 */
class IdentityProvider extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["attributeMapping"] = state ? state.attributeMapping : undefined;
            inputs["idpIdentifiers"] = state ? state.idpIdentifiers : undefined;
            inputs["providerDetails"] = state ? state.providerDetails : undefined;
            inputs["providerName"] = state ? state.providerName : undefined;
            inputs["providerType"] = state ? state.providerType : undefined;
            inputs["userPoolId"] = state ? state.userPoolId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.providerDetails === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'providerDetails'");
            }
            if ((!args || args.providerName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'providerName'");
            }
            if ((!args || args.providerType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'providerType'");
            }
            if ((!args || args.userPoolId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'userPoolId'");
            }
            inputs["attributeMapping"] = args ? args.attributeMapping : undefined;
            inputs["idpIdentifiers"] = args ? args.idpIdentifiers : undefined;
            inputs["providerDetails"] = args ? args.providerDetails : undefined;
            inputs["providerName"] = args ? args.providerName : undefined;
            inputs["providerType"] = args ? args.providerType : undefined;
            inputs["userPoolId"] = args ? args.userPoolId : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(IdentityProvider.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing IdentityProvider resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new IdentityProvider(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of IdentityProvider.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === IdentityProvider.__pulumiType;
    }
}
exports.IdentityProvider = IdentityProvider;
/** @internal */
IdentityProvider.__pulumiType = 'aws:cognito/identityProvider:IdentityProvider';
