// @pulumi/aws/cognito/resourceServer.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Cognito Resource Server.
 *
 * ## Example Usage
 * ### Create a basic resource server
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const pool = new aws.cognito.UserPool("pool", {});
 * const resource = new aws.cognito.ResourceServer("resource", {
 *     identifier: "https://example.com",
 *     userPoolId: pool.id,
 * });
 * ```
 * ### Create a resource server with sample-scope
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const pool = new aws.cognito.UserPool("pool", {});
 * const resource = new aws.cognito.ResourceServer("resource", {
 *     identifier: "https://example.com",
 *     scopes: [{
 *         scopeName: "sample-scope",
 *         scopeDescription: "a Sample Scope Description",
 *     }],
 *     userPoolId: pool.id,
 * });
 * ```
 *
 * ## Import
 *
 * `aws_cognito_resource_server` can be imported using their User Pool ID and Identifier, e.g.
 *
 * ```sh
 *  $ pulumi import aws:cognito/resourceServer:ResourceServer example xxx_yyyyy|https://example.com
 * ```
 */
class ResourceServer extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["identifier"] = state ? state.identifier : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["scopeIdentifiers"] = state ? state.scopeIdentifiers : undefined;
            inputs["scopes"] = state ? state.scopes : undefined;
            inputs["userPoolId"] = state ? state.userPoolId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.identifier === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'identifier'");
            }
            if ((!args || args.userPoolId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'userPoolId'");
            }
            inputs["identifier"] = args ? args.identifier : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["scopes"] = args ? args.scopes : undefined;
            inputs["userPoolId"] = args ? args.userPoolId : undefined;
            inputs["scopeIdentifiers"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(ResourceServer.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ResourceServer resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ResourceServer(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResourceServer.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResourceServer.__pulumiType;
    }
}
exports.ResourceServer = ResourceServer;
/** @internal */
ResourceServer.__pulumiType = 'aws:cognito/resourceServer:ResourceServer';
