// @pulumi/aws/cognito/userPoolDomain.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Cognito User Pool Domain resource.
 *
 * ## Example Usage
 * ### Amazon Cognito domain
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.cognito.UserPool("example", {});
 * const main = new aws.cognito.UserPoolDomain("main", {
 *     domain: "example-domain",
 *     userPoolId: example.id,
 * });
 * ```
 * ### Custom Cognito domain
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const exampleUserPool = new aws.cognito.UserPool("exampleUserPool", {});
 * const main = new aws.cognito.UserPoolDomain("main", {
 *     domain: "example-domain.example.com",
 *     certificateArn: aws_acm_certificate.cert.arn,
 *     userPoolId: exampleUserPool.id,
 * });
 * const exampleZone = aws.route53.getZone({
 *     name: "example.com",
 * });
 * const auth_cognito_A = new aws.route53.Record("auth-cognito-A", {
 *     name: main.domain,
 *     type: "A",
 *     zoneId: exampleZone.then(exampleZone => exampleZone.zoneId),
 *     aliases: [{
 *         evaluateTargetHealth: false,
 *         name: main.cloudfrontDistributionArn,
 *         zoneId: "Z2FDTNDATAQYW2",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * Cognito User Pool Domains can be imported using the `domain`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:cognito/userPoolDomain:UserPoolDomain main <domain>
 * ```
 */
class UserPoolDomain extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["awsAccountId"] = state ? state.awsAccountId : undefined;
            inputs["certificateArn"] = state ? state.certificateArn : undefined;
            inputs["cloudfrontDistributionArn"] = state ? state.cloudfrontDistributionArn : undefined;
            inputs["domain"] = state ? state.domain : undefined;
            inputs["s3Bucket"] = state ? state.s3Bucket : undefined;
            inputs["userPoolId"] = state ? state.userPoolId : undefined;
            inputs["version"] = state ? state.version : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.domain === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'domain'");
            }
            if ((!args || args.userPoolId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'userPoolId'");
            }
            inputs["certificateArn"] = args ? args.certificateArn : undefined;
            inputs["domain"] = args ? args.domain : undefined;
            inputs["userPoolId"] = args ? args.userPoolId : undefined;
            inputs["awsAccountId"] = undefined /*out*/;
            inputs["cloudfrontDistributionArn"] = undefined /*out*/;
            inputs["s3Bucket"] = undefined /*out*/;
            inputs["version"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(UserPoolDomain.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing UserPoolDomain resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new UserPoolDomain(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of UserPoolDomain.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === UserPoolDomain.__pulumiType;
    }
}
exports.UserPoolDomain = UserPoolDomain;
/** @internal */
UserPoolDomain.__pulumiType = 'aws:cognito/userPoolDomain:UserPoolDomain';
