// @pulumi/aws/s3/cannedAcl.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
/** Use CannedAcl.Private instead. */
exports.PrivateAcl = "private";
/** Use CannedAcl.PublicRead instead. */
exports.PublicReadAcl = "public-read";
/** Use CannedAcl.PublicReadWrite instead. */
exports.PublicReadWriteAcl = "public-read-write";
/** Use CannedAcl.AwsExecRead instead. */
exports.AwsExecReadAcl = "aws-exec-read";
/** Use CannedAcl.AuthenticatedRead instead. */
exports.AuthenticatedReadAcl = "authenticated-read";
/** Use CannedAcl.BucketOwnerRead instead. */
exports.BucketOwnerReadAcl = "bucket-owner-read";
/** Use CannedAcl.BucketOwnerFullControl instead. */
exports.BucketOwnerFullControlAcl = "bucket-owner-full-control";
/** Use CannedAcl.LogDeliveryWrite instead. */
exports.LogDeliveryWriteAcl = "log-delivery-write";
