// @pulumi/aws/lambda/codeSigningConfig.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Lambda Code Signing Config resource. A code signing configuration defines a list of allowed signing profiles and defines the code-signing validation policy (action to be taken if deployment validation checks fail).
 *
 * For information about Lambda code signing configurations and how to use them, see [configuring code signing for Lambda functions](https://docs.aws.amazon.com/lambda/latest/dg/configuration-codesigning.html)
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const newCsc = new aws.lambda.CodeSigningConfig("newCsc", {
 *     allowedPublishers: {
 *         signingProfileVersionArns: [
 *             aws_signer_signing_profile.example1.arn,
 *             aws_signer_signing_profile.example2.arn,
 *         ],
 *     },
 *     policies: {
 *         untrustedArtifactOnDeployment: "Warn",
 *     },
 *     description: "My awesome code signing config.",
 * });
 * ```
 *
 * ## Import
 *
 * Code Signing Configs can be imported using their ARN, e.g.
 *
 * ```sh
 *  $ pulumi import aws:lambda/codeSigningConfig:CodeSigningConfig imported_csc arn:aws:lambda:us-west-2:123456789012:code-signing-config:csc-0f6c334abcdea4d8b
 * ```
 */
class CodeSigningConfig extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["allowedPublishers"] = state ? state.allowedPublishers : undefined;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["configId"] = state ? state.configId : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["lastModified"] = state ? state.lastModified : undefined;
            inputs["policies"] = state ? state.policies : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.allowedPublishers === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'allowedPublishers'");
            }
            inputs["allowedPublishers"] = args ? args.allowedPublishers : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["policies"] = args ? args.policies : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["configId"] = undefined /*out*/;
            inputs["lastModified"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(CodeSigningConfig.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing CodeSigningConfig resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new CodeSigningConfig(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of CodeSigningConfig.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === CodeSigningConfig.__pulumiType;
    }
}
exports.CodeSigningConfig = CodeSigningConfig;
/** @internal */
CodeSigningConfig.__pulumiType = 'aws:lambda/codeSigningConfig:CodeSigningConfig';
