// @pulumi/aws/storagegateway/tapePool.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an AWS Storage Gateway Tape Pool.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.storagegateway.TapePool("example", {
 *     poolName: "example",
 *     storageClass: "GLACIER",
 * });
 * ```
 *
 * ## Import
 *
 * `aws_storagegateway_tape_pool` can be imported by using the volume Amazon Resource Name (ARN), e.g.
 *
 * ```sh
 *  $ pulumi import aws:storagegateway/tapePool:TapePool example arn:aws:storagegateway:us-east-1:123456789012:tapepool/pool-12345678
 * ```
 */
class TapePool extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["poolName"] = state ? state.poolName : undefined;
            inputs["retentionLockTimeInDays"] = state ? state.retentionLockTimeInDays : undefined;
            inputs["retentionLockType"] = state ? state.retentionLockType : undefined;
            inputs["storageClass"] = state ? state.storageClass : undefined;
            inputs["tags"] = state ? state.tags : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.poolName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'poolName'");
            }
            if ((!args || args.storageClass === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'storageClass'");
            }
            inputs["poolName"] = args ? args.poolName : undefined;
            inputs["retentionLockTimeInDays"] = args ? args.retentionLockTimeInDays : undefined;
            inputs["retentionLockType"] = args ? args.retentionLockType : undefined;
            inputs["storageClass"] = args ? args.storageClass : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(TapePool.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing TapePool resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new TapePool(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of TapePool.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === TapePool.__pulumiType;
    }
}
exports.TapePool = TapePool;
/** @internal */
TapePool.__pulumiType = 'aws:storagegateway/tapePool:TapePool';
