// @pulumi/aws/apigatewayv2/integration.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an Amazon API Gateway Version 2 integration.
 * More information can be found in the [Amazon API Gateway Developer Guide](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api.html).
 *
 * ## Example Usage
 * ### Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.apigatewayv2.Integration("example", {
 *     apiId: aws_apigatewayv2_api.example.id,
 *     integrationType: "MOCK",
 * });
 * ```
 * ### Lambda Integration
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const exampleFunction = new aws.lambda.Function("exampleFunction", {
 *     code: new pulumi.asset.FileArchive("example.zip"),
 *     role: aws_iam_role.example.arn,
 *     handler: "index.handler",
 *     runtime: "nodejs12.x",
 * });
 * const exampleIntegration = new aws.apigatewayv2.Integration("exampleIntegration", {
 *     apiId: aws_apigatewayv2_api.example.id,
 *     integrationType: "AWS",
 *     connectionType: "INTERNET",
 *     contentHandlingStrategy: "CONVERT_TO_TEXT",
 *     description: "Lambda example",
 *     integrationMethod: "POST",
 *     integrationUri: exampleFunction.invokeArn,
 *     passthroughBehavior: "WHEN_NO_MATCH",
 * });
 * ```
 * ### AWS Service Integration
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.apigatewayv2.Integration("example", {
 *     apiId: aws_apigatewayv2_api.example.id,
 *     credentialsArn: aws_iam_role.example.arn,
 *     description: "SQS example",
 *     integrationType: "AWS_PROXY",
 *     integrationSubtype: "SQS-SendMessage",
 *     requestParameters: {
 *         QueueUrl: `$request.header.queueUrl`,
 *         MessageBody: `$request.body.message`,
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * `aws_apigatewayv2_integration` can be imported by using the API identifier and integration identifier, e.g.
 *
 * ```sh
 *  $ pulumi import aws:apigatewayv2/integration:Integration example aabbccddee/1122334
 * ```
 */
class Integration extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["apiId"] = state ? state.apiId : undefined;
            inputs["connectionId"] = state ? state.connectionId : undefined;
            inputs["connectionType"] = state ? state.connectionType : undefined;
            inputs["contentHandlingStrategy"] = state ? state.contentHandlingStrategy : undefined;
            inputs["credentialsArn"] = state ? state.credentialsArn : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["integrationMethod"] = state ? state.integrationMethod : undefined;
            inputs["integrationResponseSelectionExpression"] = state ? state.integrationResponseSelectionExpression : undefined;
            inputs["integrationSubtype"] = state ? state.integrationSubtype : undefined;
            inputs["integrationType"] = state ? state.integrationType : undefined;
            inputs["integrationUri"] = state ? state.integrationUri : undefined;
            inputs["passthroughBehavior"] = state ? state.passthroughBehavior : undefined;
            inputs["payloadFormatVersion"] = state ? state.payloadFormatVersion : undefined;
            inputs["requestParameters"] = state ? state.requestParameters : undefined;
            inputs["requestTemplates"] = state ? state.requestTemplates : undefined;
            inputs["templateSelectionExpression"] = state ? state.templateSelectionExpression : undefined;
            inputs["timeoutMilliseconds"] = state ? state.timeoutMilliseconds : undefined;
            inputs["tlsConfig"] = state ? state.tlsConfig : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.apiId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'apiId'");
            }
            if ((!args || args.integrationType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'integrationType'");
            }
            inputs["apiId"] = args ? args.apiId : undefined;
            inputs["connectionId"] = args ? args.connectionId : undefined;
            inputs["connectionType"] = args ? args.connectionType : undefined;
            inputs["contentHandlingStrategy"] = args ? args.contentHandlingStrategy : undefined;
            inputs["credentialsArn"] = args ? args.credentialsArn : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["integrationMethod"] = args ? args.integrationMethod : undefined;
            inputs["integrationSubtype"] = args ? args.integrationSubtype : undefined;
            inputs["integrationType"] = args ? args.integrationType : undefined;
            inputs["integrationUri"] = args ? args.integrationUri : undefined;
            inputs["passthroughBehavior"] = args ? args.passthroughBehavior : undefined;
            inputs["payloadFormatVersion"] = args ? args.payloadFormatVersion : undefined;
            inputs["requestParameters"] = args ? args.requestParameters : undefined;
            inputs["requestTemplates"] = args ? args.requestTemplates : undefined;
            inputs["templateSelectionExpression"] = args ? args.templateSelectionExpression : undefined;
            inputs["timeoutMilliseconds"] = args ? args.timeoutMilliseconds : undefined;
            inputs["tlsConfig"] = args ? args.tlsConfig : undefined;
            inputs["integrationResponseSelectionExpression"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Integration.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Integration resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Integration(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Integration.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Integration.__pulumiType;
    }
}
exports.Integration = Integration;
/** @internal */
Integration.__pulumiType = 'aws:apigatewayv2/integration:Integration';
