// @pulumi/aws/apigatewayv2/routeResponse.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an Amazon API Gateway Version 2 route response.
 * More information can be found in the [Amazon API Gateway Developer Guide](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api.html).
 *
 * ## Example Usage
 * ### Basic
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.apigatewayv2.RouteResponse("example", {
 *     apiId: aws_apigatewayv2_api.example.id,
 *     routeId: aws_apigatewayv2_route.example.id,
 *     routeResponseKey: `$default`,
 * });
 * ```
 *
 * ## Import
 *
 * `aws_apigatewayv2_route_response` can be imported by using the API identifier, route identifier and route response identifier, e.g.
 *
 * ```sh
 *  $ pulumi import aws:apigatewayv2/routeResponse:RouteResponse example aabbccddee/1122334/998877
 * ```
 */
class RouteResponse extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["apiId"] = state ? state.apiId : undefined;
            inputs["modelSelectionExpression"] = state ? state.modelSelectionExpression : undefined;
            inputs["responseModels"] = state ? state.responseModels : undefined;
            inputs["routeId"] = state ? state.routeId : undefined;
            inputs["routeResponseKey"] = state ? state.routeResponseKey : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.apiId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'apiId'");
            }
            if ((!args || args.routeId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'routeId'");
            }
            if ((!args || args.routeResponseKey === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'routeResponseKey'");
            }
            inputs["apiId"] = args ? args.apiId : undefined;
            inputs["modelSelectionExpression"] = args ? args.modelSelectionExpression : undefined;
            inputs["responseModels"] = args ? args.responseModels : undefined;
            inputs["routeId"] = args ? args.routeId : undefined;
            inputs["routeResponseKey"] = args ? args.routeResponseKey : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(RouteResponse.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing RouteResponse resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new RouteResponse(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of RouteResponse.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === RouteResponse.__pulumiType;
    }
}
exports.RouteResponse = RouteResponse;
/** @internal */
RouteResponse.__pulumiType = 'aws:apigatewayv2/routeResponse:RouteResponse';
