// @pulumi/aws/iam/openIdConnectProvider.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an IAM OpenID Connect provider.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const defaultOpenIdConnectProvider = new aws.iam.OpenIdConnectProvider("default", {
 *     clientIdLists: ["266362248691-342342xasdasdasda-apps.googleusercontent.com"],
 *     thumbprintLists: [],
 *     url: "https://accounts.google.com",
 * });
 * ```
 *
 * ## Import
 *
 * IAM OpenID Connect Providers can be imported using the `arn`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:iam/openIdConnectProvider:OpenIdConnectProvider default arn:aws:iam::123456789012:oidc-provider/accounts.google.com
 * ```
 */
class OpenIdConnectProvider extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["clientIdLists"] = state ? state.clientIdLists : undefined;
            inputs["thumbprintLists"] = state ? state.thumbprintLists : undefined;
            inputs["url"] = state ? state.url : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.clientIdLists === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'clientIdLists'");
            }
            if ((!args || args.thumbprintLists === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'thumbprintLists'");
            }
            if ((!args || args.url === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'url'");
            }
            inputs["clientIdLists"] = args ? args.clientIdLists : undefined;
            inputs["thumbprintLists"] = args ? args.thumbprintLists : undefined;
            inputs["url"] = args ? args.url : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(OpenIdConnectProvider.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing OpenIdConnectProvider resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new OpenIdConnectProvider(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of OpenIdConnectProvider.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === OpenIdConnectProvider.__pulumiType;
    }
}
exports.OpenIdConnectProvider = OpenIdConnectProvider;
/** @internal */
OpenIdConnectProvider.__pulumiType = 'aws:iam/openIdConnectProvider:OpenIdConnectProvider';
