// @pulumi/aws/macie/s3bucketAssociation.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * > **NOTE:** This resource interacts with [Amazon Macie Classic](https://docs.aws.amazon.com/macie/latest/userguide/what-is-macie.html). Macie Classic cannot be activated in new accounts. See the [FAQ](https://aws.amazon.com/macie/classic-faqs/) for more details.
 *
 * Associates an S3 resource with Amazon Macie for monitoring and data classification.
 *
 * > **NOTE:** Before using Amazon Macie for the first time it must be enabled manually. Instructions are [here](https://docs.aws.amazon.com/macie/latest/userguide/macie-setting-up.html#macie-setting-up-enable).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.macie.S3BucketAssociation("example", {
 *     bucketName: "tf-macie-example",
 *     classificationType: {
 *         oneTime: "FULL",
 *     },
 *     prefix: "data",
 * });
 * ```
 */
class S3BucketAssociation extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["bucketName"] = state ? state.bucketName : undefined;
            inputs["classificationType"] = state ? state.classificationType : undefined;
            inputs["memberAccountId"] = state ? state.memberAccountId : undefined;
            inputs["prefix"] = state ? state.prefix : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.bucketName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'bucketName'");
            }
            inputs["bucketName"] = args ? args.bucketName : undefined;
            inputs["classificationType"] = args ? args.classificationType : undefined;
            inputs["memberAccountId"] = args ? args.memberAccountId : undefined;
            inputs["prefix"] = args ? args.prefix : undefined;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(S3BucketAssociation.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing S3BucketAssociation resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new S3BucketAssociation(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of S3BucketAssociation.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === S3BucketAssociation.__pulumiType;
    }
}
exports.S3BucketAssociation = S3BucketAssociation;
/** @internal */
S3BucketAssociation.__pulumiType = 'aws:macie/s3BucketAssociation:S3BucketAssociation';
