// @pulumi/aws/imagebuilder/imagePipeline.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an Image Builder Image Pipeline.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.imagebuilder.ImagePipeline("example", {
 *     imageRecipeArn: aws_imagebuilder_image_recipe.example.arn,
 *     infrastructureConfigurationArn: aws_imagebuilder_infrastructure_configuration.example.arn,
 *     schedule: {
 *         scheduleExpression: "cron(0 0 * * *)",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * `aws_imagebuilder_image_pipeline` resources can be imported using the Amazon Resource Name (ARN), e.g.
 *
 * ```sh
 *  $ pulumi import aws:imagebuilder/imagePipeline:ImagePipeline example arn:aws:imagebuilder:us-east-1:123456789012:image-pipeline/example
 * ```
 */
class ImagePipeline extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["dateCreated"] = state ? state.dateCreated : undefined;
            inputs["dateLastRun"] = state ? state.dateLastRun : undefined;
            inputs["dateNextRun"] = state ? state.dateNextRun : undefined;
            inputs["dateUpdated"] = state ? state.dateUpdated : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["distributionConfigurationArn"] = state ? state.distributionConfigurationArn : undefined;
            inputs["enhancedImageMetadataEnabled"] = state ? state.enhancedImageMetadataEnabled : undefined;
            inputs["imageRecipeArn"] = state ? state.imageRecipeArn : undefined;
            inputs["imageTestsConfiguration"] = state ? state.imageTestsConfiguration : undefined;
            inputs["infrastructureConfigurationArn"] = state ? state.infrastructureConfigurationArn : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["platform"] = state ? state.platform : undefined;
            inputs["schedule"] = state ? state.schedule : undefined;
            inputs["status"] = state ? state.status : undefined;
            inputs["tags"] = state ? state.tags : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.imageRecipeArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'imageRecipeArn'");
            }
            if ((!args || args.infrastructureConfigurationArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'infrastructureConfigurationArn'");
            }
            inputs["description"] = args ? args.description : undefined;
            inputs["distributionConfigurationArn"] = args ? args.distributionConfigurationArn : undefined;
            inputs["enhancedImageMetadataEnabled"] = args ? args.enhancedImageMetadataEnabled : undefined;
            inputs["imageRecipeArn"] = args ? args.imageRecipeArn : undefined;
            inputs["imageTestsConfiguration"] = args ? args.imageTestsConfiguration : undefined;
            inputs["infrastructureConfigurationArn"] = args ? args.infrastructureConfigurationArn : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["schedule"] = args ? args.schedule : undefined;
            inputs["status"] = args ? args.status : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["dateCreated"] = undefined /*out*/;
            inputs["dateLastRun"] = undefined /*out*/;
            inputs["dateNextRun"] = undefined /*out*/;
            inputs["dateUpdated"] = undefined /*out*/;
            inputs["platform"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(ImagePipeline.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ImagePipeline resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ImagePipeline(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ImagePipeline.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ImagePipeline.__pulumiType;
    }
}
exports.ImagePipeline = ImagePipeline;
/** @internal */
ImagePipeline.__pulumiType = 'aws:imagebuilder/imagePipeline:ImagePipeline';
