// @pulumi/aws/imagebuilder/infrastructureConfiguration.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an Image Builder Infrastructure Configuration.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.imagebuilder.InfrastructureConfiguration("example", {
 *     description: "example description",
 *     instanceProfileName: aws_iam_instance_profile.example.name,
 *     instanceTypes: [
 *         "t2.nano",
 *         "t3.micro",
 *     ],
 *     keyPair: aws_key_pair.example.key_name,
 *     securityGroupIds: [aws_security_group.example.id],
 *     snsTopicArn: aws_sns_topic.example.arn,
 *     subnetId: aws_subnet.main.id,
 *     terminateInstanceOnFailure: true,
 *     logging: {
 *         s3Logs: {
 *             s3BucketName: aws_s3_bucket.example.bucket,
 *             s3KeyPrefix: "logs",
 *         },
 *     },
 *     tags: {
 *         foo: "bar",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * `aws_imagebuilder_infrastructure_configuration` can be imported using the Amazon Resource Name (ARN), e.g.
 *
 * ```sh
 *  $ pulumi import aws:imagebuilder/infrastructureConfiguration:InfrastructureConfiguration example arn:aws:imagebuilder:us-east-1:123456789012:infrastructure-component/example
 * ```
 */
class InfrastructureConfiguration extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["dateCreated"] = state ? state.dateCreated : undefined;
            inputs["dateUpdated"] = state ? state.dateUpdated : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["instanceProfileName"] = state ? state.instanceProfileName : undefined;
            inputs["instanceTypes"] = state ? state.instanceTypes : undefined;
            inputs["keyPair"] = state ? state.keyPair : undefined;
            inputs["logging"] = state ? state.logging : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["resourceTags"] = state ? state.resourceTags : undefined;
            inputs["securityGroupIds"] = state ? state.securityGroupIds : undefined;
            inputs["snsTopicArn"] = state ? state.snsTopicArn : undefined;
            inputs["subnetId"] = state ? state.subnetId : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["terminateInstanceOnFailure"] = state ? state.terminateInstanceOnFailure : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.instanceProfileName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'instanceProfileName'");
            }
            inputs["description"] = args ? args.description : undefined;
            inputs["instanceProfileName"] = args ? args.instanceProfileName : undefined;
            inputs["instanceTypes"] = args ? args.instanceTypes : undefined;
            inputs["keyPair"] = args ? args.keyPair : undefined;
            inputs["logging"] = args ? args.logging : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["resourceTags"] = args ? args.resourceTags : undefined;
            inputs["securityGroupIds"] = args ? args.securityGroupIds : undefined;
            inputs["snsTopicArn"] = args ? args.snsTopicArn : undefined;
            inputs["subnetId"] = args ? args.subnetId : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["terminateInstanceOnFailure"] = args ? args.terminateInstanceOnFailure : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["dateCreated"] = undefined /*out*/;
            inputs["dateUpdated"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(InfrastructureConfiguration.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing InfrastructureConfiguration resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new InfrastructureConfiguration(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of InfrastructureConfiguration.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === InfrastructureConfiguration.__pulumiType;
    }
}
exports.InfrastructureConfiguration = InfrastructureConfiguration;
/** @internal */
InfrastructureConfiguration.__pulumiType = 'aws:imagebuilder/infrastructureConfiguration:InfrastructureConfiguration';
