// @pulumi/aws/globalaccelerator/endpointGroup.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Global Accelerator endpoint group.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.globalaccelerator.EndpointGroup("example", {
 *     listenerArn: aws_globalaccelerator_listener.example.id,
 *     endpointConfigurations: [{
 *         endpointId: aws_lb.example.arn,
 *         weight: 100,
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * Global Accelerator endpoint groups can be imported using the `id`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:globalaccelerator/endpointGroup:EndpointGroup example arn:aws:globalaccelerator::111111111111:accelerator/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/listener/xxxxxxx/endpoint-group/xxxxxxxx
 * ```
 */
class EndpointGroup extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["endpointConfigurations"] = state ? state.endpointConfigurations : undefined;
            inputs["endpointGroupRegion"] = state ? state.endpointGroupRegion : undefined;
            inputs["healthCheckIntervalSeconds"] = state ? state.healthCheckIntervalSeconds : undefined;
            inputs["healthCheckPath"] = state ? state.healthCheckPath : undefined;
            inputs["healthCheckPort"] = state ? state.healthCheckPort : undefined;
            inputs["healthCheckProtocol"] = state ? state.healthCheckProtocol : undefined;
            inputs["listenerArn"] = state ? state.listenerArn : undefined;
            inputs["portOverrides"] = state ? state.portOverrides : undefined;
            inputs["thresholdCount"] = state ? state.thresholdCount : undefined;
            inputs["trafficDialPercentage"] = state ? state.trafficDialPercentage : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.listenerArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'listenerArn'");
            }
            inputs["endpointConfigurations"] = args ? args.endpointConfigurations : undefined;
            inputs["endpointGroupRegion"] = args ? args.endpointGroupRegion : undefined;
            inputs["healthCheckIntervalSeconds"] = args ? args.healthCheckIntervalSeconds : undefined;
            inputs["healthCheckPath"] = args ? args.healthCheckPath : undefined;
            inputs["healthCheckPort"] = args ? args.healthCheckPort : undefined;
            inputs["healthCheckProtocol"] = args ? args.healthCheckProtocol : undefined;
            inputs["listenerArn"] = args ? args.listenerArn : undefined;
            inputs["portOverrides"] = args ? args.portOverrides : undefined;
            inputs["thresholdCount"] = args ? args.thresholdCount : undefined;
            inputs["trafficDialPercentage"] = args ? args.trafficDialPercentage : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(EndpointGroup.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing EndpointGroup resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new EndpointGroup(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of EndpointGroup.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EndpointGroup.__pulumiType;
    }
}
exports.EndpointGroup = EndpointGroup;
/** @internal */
EndpointGroup.__pulumiType = 'aws:globalaccelerator/endpointGroup:EndpointGroup';
