// @pulumi/aws/ssm/activation.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Registers an on-premises server or virtual machine with Amazon EC2 so that it can be managed using Run Command.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const testRole = new aws.iam.Role("testRole", {assumeRolePolicy: `  {
 *     "Version": "2012-10-17",
 *     "Statement": {
 *       "Effect": "Allow",
 *       "Principal": {"Service": "ssm.amazonaws.com"},
 *       "Action": "sts:AssumeRole"
 *     }
 *   }
 * `});
 * const testAttach = new aws.iam.RolePolicyAttachment("testAttach", {
 *     role: testRole.name,
 *     policyArn: "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
 * });
 * const foo = new aws.ssm.Activation("foo", {
 *     description: "Test",
 *     iamRole: testRole.id,
 *     registrationLimit: "5",
 * }, {
 *     dependsOn: [testAttach],
 * });
 * ```
 *
 * ## Import
 *
 * AWS SSM Activation can be imported using the `id`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:ssm/activation:Activation example e488f2f6-e686-4afb-8a04-ef6dfEXAMPLE
 * ```
 */
class Activation extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["activationCode"] = state ? state.activationCode : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["expirationDate"] = state ? state.expirationDate : undefined;
            inputs["expired"] = state ? state.expired : undefined;
            inputs["iamRole"] = state ? state.iamRole : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["registrationCount"] = state ? state.registrationCount : undefined;
            inputs["registrationLimit"] = state ? state.registrationLimit : undefined;
            inputs["tags"] = state ? state.tags : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.iamRole === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'iamRole'");
            }
            inputs["description"] = args ? args.description : undefined;
            inputs["expirationDate"] = args ? args.expirationDate : undefined;
            inputs["iamRole"] = args ? args.iamRole : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["registrationLimit"] = args ? args.registrationLimit : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["activationCode"] = undefined /*out*/;
            inputs["expired"] = undefined /*out*/;
            inputs["registrationCount"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Activation.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Activation resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Activation(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Activation.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Activation.__pulumiType;
    }
}
exports.Activation = Activation;
/** @internal */
Activation.__pulumiType = 'aws:ssm/activation:Activation';
