// @pulumi/aws/msk/configuration.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an Amazon Managed Streaming for Kafka configuration. More information can be found on the [MSK Developer Guide](https://docs.aws.amazon.com/msk/latest/developerguide/msk-configuration.html).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.msk.Configuration("example", {
 *     kafkaVersions: ["2.1.0"],
 *     serverProperties: `auto.create.topics.enable = true
 * delete.topic.enable = true
 * `,
 * });
 * ```
 *
 * ## Import
 *
 * MSK configurations can be imported using the configuration ARN, e.g.
 *
 * ```sh
 *  $ pulumi import aws:msk/configuration:Configuration example arn:aws:kafka:us-west-2:123456789012:configuration/example/279c0212-d057-4dba-9aa9-1c4e5a25bfc7-3
 * ```
 */
class Configuration extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["kafkaVersions"] = state ? state.kafkaVersions : undefined;
            inputs["latestRevision"] = state ? state.latestRevision : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["serverProperties"] = state ? state.serverProperties : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.kafkaVersions === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'kafkaVersions'");
            }
            if ((!args || args.serverProperties === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'serverProperties'");
            }
            inputs["description"] = args ? args.description : undefined;
            inputs["kafkaVersions"] = args ? args.kafkaVersions : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["serverProperties"] = args ? args.serverProperties : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["latestRevision"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Configuration.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Configuration resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Configuration(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Configuration.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Configuration.__pulumiType;
    }
}
exports.Configuration = Configuration;
/** @internal */
Configuration.__pulumiType = 'aws:msk/configuration:Configuration';
