// @pulumi/aws/autoscaling/metrics.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
/** Use MetricsGranularity.OneMinute instead. */
exports.OneMinuteMetricsGranularity = "1Minute";
/** Use Metric.GroupMinSize instead. */
exports.GroupMinSizeMetric = "GroupMinSize";
/** Use Metric.GroupMaxSize instead. */
exports.GroupMaxSizeMetric = "GroupMaxSize";
/** Use Metric.GroupDesiredCapacity instead. */
exports.GroupDesiredCapacityMetric = "GroupDesiredCapacity";
/** Use Metric.GroupGroupInServiceInstancesMinSize instead. */
exports.GroupInServiceInstancesMetric = "GroupInServiceInstances";
/** Use Metric.GroupPendingInstances instead. */
exports.GroupPendingInstances = "GroupPendingInstances";
/** Use Metric.GroupStandbyInstances instead. */
exports.GroupStandbyInstances = "GroupStandbyInstances";
/** Use Metric.GroupTerminatingInstances instead. */
exports.GroupTerminatingInstances = "GroupTerminatingInstances";
/** Use Metric.GroupTotalInstances instead. */
exports.GroupTotalInstances = "GroupTotalInstances";
