// @pulumi/aws/ec2transitgateway/transitGateway.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an EC2 Transit Gateway.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.ec2transitgateway.TransitGateway("example", {
 *     description: "example",
 * });
 * ```
 *
 * ## Import
 *
 * `aws_ec2_transit_gateway` can be imported by using the EC2 Transit Gateway identifier, e.g.
 *
 * ```sh
 *  $ pulumi import aws:ec2transitgateway/transitGateway:TransitGateway example tgw-12345678
 * ```
 */
class TransitGateway extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["amazonSideAsn"] = state ? state.amazonSideAsn : undefined;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["associationDefaultRouteTableId"] = state ? state.associationDefaultRouteTableId : undefined;
            inputs["autoAcceptSharedAttachments"] = state ? state.autoAcceptSharedAttachments : undefined;
            inputs["defaultRouteTableAssociation"] = state ? state.defaultRouteTableAssociation : undefined;
            inputs["defaultRouteTablePropagation"] = state ? state.defaultRouteTablePropagation : undefined;
            inputs["description"] = state ? state.description : undefined;
            inputs["dnsSupport"] = state ? state.dnsSupport : undefined;
            inputs["ownerId"] = state ? state.ownerId : undefined;
            inputs["propagationDefaultRouteTableId"] = state ? state.propagationDefaultRouteTableId : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["vpnEcmpSupport"] = state ? state.vpnEcmpSupport : undefined;
        }
        else {
            const args = argsOrState;
            inputs["amazonSideAsn"] = args ? args.amazonSideAsn : undefined;
            inputs["autoAcceptSharedAttachments"] = args ? args.autoAcceptSharedAttachments : undefined;
            inputs["defaultRouteTableAssociation"] = args ? args.defaultRouteTableAssociation : undefined;
            inputs["defaultRouteTablePropagation"] = args ? args.defaultRouteTablePropagation : undefined;
            inputs["description"] = args ? args.description : undefined;
            inputs["dnsSupport"] = args ? args.dnsSupport : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["vpnEcmpSupport"] = args ? args.vpnEcmpSupport : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["associationDefaultRouteTableId"] = undefined /*out*/;
            inputs["ownerId"] = undefined /*out*/;
            inputs["propagationDefaultRouteTableId"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(TransitGateway.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing TransitGateway resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new TransitGateway(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of TransitGateway.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === TransitGateway.__pulumiType;
    }
}
exports.TransitGateway = TransitGateway;
/** @internal */
TransitGateway.__pulumiType = 'aws:ec2transitgateway/transitGateway:TransitGateway';
