// @pulumi/aws/rds/proxy.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an RDS DB proxy resource. For additional information, see the [RDS User Guide](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy.html).
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.rds.Proxy("example", {
 *     debugLogging: false,
 *     engineFamily: "MYSQL",
 *     idleClientTimeout: 1800,
 *     requireTls: true,
 *     roleArn: aws_iam_role.example.arn,
 *     vpcSecurityGroupIds: [aws_security_group.example.id],
 *     vpcSubnetIds: [aws_subnet.example.id],
 *     auths: [{
 *         authScheme: "SECRETS",
 *         description: "example",
 *         iamAuth: "DISABLED",
 *         secretArn: aws_secretsmanager_secret.example.arn,
 *     }],
 *     tags: {
 *         Name: "example",
 *         Key: "value",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * DB proxies can be imported using the `name`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:rds/proxy:Proxy example example
 * ```
 */
class Proxy extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["auths"] = state ? state.auths : undefined;
            inputs["debugLogging"] = state ? state.debugLogging : undefined;
            inputs["endpoint"] = state ? state.endpoint : undefined;
            inputs["engineFamily"] = state ? state.engineFamily : undefined;
            inputs["idleClientTimeout"] = state ? state.idleClientTimeout : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["requireTls"] = state ? state.requireTls : undefined;
            inputs["roleArn"] = state ? state.roleArn : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["vpcSecurityGroupIds"] = state ? state.vpcSecurityGroupIds : undefined;
            inputs["vpcSubnetIds"] = state ? state.vpcSubnetIds : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.auths === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'auths'");
            }
            if ((!args || args.engineFamily === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'engineFamily'");
            }
            if ((!args || args.roleArn === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'roleArn'");
            }
            if ((!args || args.vpcSubnetIds === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'vpcSubnetIds'");
            }
            inputs["auths"] = args ? args.auths : undefined;
            inputs["debugLogging"] = args ? args.debugLogging : undefined;
            inputs["engineFamily"] = args ? args.engineFamily : undefined;
            inputs["idleClientTimeout"] = args ? args.idleClientTimeout : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["requireTls"] = args ? args.requireTls : undefined;
            inputs["roleArn"] = args ? args.roleArn : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["vpcSecurityGroupIds"] = args ? args.vpcSecurityGroupIds : undefined;
            inputs["vpcSubnetIds"] = args ? args.vpcSubnetIds : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["endpoint"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Proxy.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Proxy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Proxy(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Proxy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Proxy.__pulumiType;
    }
}
exports.Proxy = Proxy;
/** @internal */
Proxy.__pulumiType = 'aws:rds/proxy:Proxy';
