// @pulumi/aws/cfg/remediationConfiguration.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an AWS Config Remediation Configuration.
 *
 * > **Note:** Config Remediation Configuration requires an existing [Config Rule](https://www.terraform.io/docs/providers/aws/r/config_config_rule.html) to be present.
 *
 * ## Example Usage
 *
 * AWS managed rules can be used by setting the source owner to `AWS` and the source identifier to the name of the managed rule. More information about AWS managed rules can be found in the [AWS Config Developer Guide](https://docs.aws.amazon.com/config/latest/developerguide/evaluate-config_use-managed-rules.html).
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const thisRule = new aws.cfg.Rule("thisRule", {source: {
 *     owner: "AWS",
 *     sourceIdentifier: "S3_BUCKET_VERSIONING_ENABLED",
 * }});
 * const thisRemediationConfiguration = new aws.cfg.RemediationConfiguration("thisRemediationConfiguration", {
 *     configRuleName: thisRule.name,
 *     resourceType: "AWS::S3::Bucket",
 *     targetType: "SSM_DOCUMENT",
 *     targetId: "AWS-EnableS3BucketEncryption",
 *     targetVersion: "1",
 *     parameters: [
 *         {
 *             name: "AutomationAssumeRole",
 *             staticValue: "arn:aws:iam::875924563244:role/security_config",
 *         },
 *         {
 *             name: "BucketName",
 *             resourceValue: "RESOURCE_ID",
 *         },
 *         {
 *             name: "SSEAlgorithm",
 *             staticValue: "AES256",
 *         },
 *     ],
 * });
 * ```
 *
 * ## Import
 *
 * Remediation Configurations can be imported using the name config_rule_name, e.g.
 *
 * ```sh
 *  $ pulumi import aws:cfg/remediationConfiguration:RemediationConfiguration this example
 * ```
 */
class RemediationConfiguration extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["configRuleName"] = state ? state.configRuleName : undefined;
            inputs["parameters"] = state ? state.parameters : undefined;
            inputs["resourceType"] = state ? state.resourceType : undefined;
            inputs["targetId"] = state ? state.targetId : undefined;
            inputs["targetType"] = state ? state.targetType : undefined;
            inputs["targetVersion"] = state ? state.targetVersion : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.configRuleName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'configRuleName'");
            }
            if ((!args || args.targetId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'targetId'");
            }
            if ((!args || args.targetType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'targetType'");
            }
            inputs["configRuleName"] = args ? args.configRuleName : undefined;
            inputs["parameters"] = args ? args.parameters : undefined;
            inputs["resourceType"] = args ? args.resourceType : undefined;
            inputs["targetId"] = args ? args.targetId : undefined;
            inputs["targetType"] = args ? args.targetType : undefined;
            inputs["targetVersion"] = args ? args.targetVersion : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(RemediationConfiguration.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing RemediationConfiguration resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new RemediationConfiguration(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of RemediationConfiguration.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === RemediationConfiguration.__pulumiType;
    }
}
exports.RemediationConfiguration = RemediationConfiguration;
/** @internal */
RemediationConfiguration.__pulumiType = 'aws:cfg/remediationConfiguration:RemediationConfiguration';
