// @pulumi/aws/directconnect/gatewayAssociationProposal.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages a Direct Connect Gateway Association Proposal, typically for enabling cross-account associations. For single account associations, see the `aws.directconnect.GatewayAssociation` resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.directconnect.GatewayAssociationProposal("example", {
 *     dxGatewayId: aws_dx_gateway.example.id,
 *     dxGatewayOwnerAccountId: aws_dx_gateway.example.owner_account_id,
 *     associatedGatewayId: aws_vpn_gateway.example.id,
 * });
 * ```
 *
 * ## Import
 *
 * Direct Connect Gateway Association Proposals can be imported using the proposal ID, e.g.
 *
 * ```sh
 *  $ pulumi import aws:directconnect/gatewayAssociationProposal:GatewayAssociationProposal example ac90e981-b718-4364-872d-65478c84fafe
 * ```
 */
class GatewayAssociationProposal extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["allowedPrefixes"] = state ? state.allowedPrefixes : undefined;
            inputs["associatedGatewayId"] = state ? state.associatedGatewayId : undefined;
            inputs["associatedGatewayOwnerAccountId"] = state ? state.associatedGatewayOwnerAccountId : undefined;
            inputs["associatedGatewayType"] = state ? state.associatedGatewayType : undefined;
            inputs["dxGatewayId"] = state ? state.dxGatewayId : undefined;
            inputs["dxGatewayOwnerAccountId"] = state ? state.dxGatewayOwnerAccountId : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.associatedGatewayId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'associatedGatewayId'");
            }
            if ((!args || args.dxGatewayId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'dxGatewayId'");
            }
            if ((!args || args.dxGatewayOwnerAccountId === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'dxGatewayOwnerAccountId'");
            }
            inputs["allowedPrefixes"] = args ? args.allowedPrefixes : undefined;
            inputs["associatedGatewayId"] = args ? args.associatedGatewayId : undefined;
            inputs["dxGatewayId"] = args ? args.dxGatewayId : undefined;
            inputs["dxGatewayOwnerAccountId"] = args ? args.dxGatewayOwnerAccountId : undefined;
            inputs["associatedGatewayOwnerAccountId"] = undefined /*out*/;
            inputs["associatedGatewayType"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(GatewayAssociationProposal.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing GatewayAssociationProposal resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new GatewayAssociationProposal(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of GatewayAssociationProposal.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === GatewayAssociationProposal.__pulumiType;
    }
}
exports.GatewayAssociationProposal = GatewayAssociationProposal;
/** @internal */
GatewayAssociationProposal.__pulumiType = 'aws:directconnect/gatewayAssociationProposal:GatewayAssociationProposal';
