// @pulumi/aws/gamelift/gameSessionQueue.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides an Gamelift Game Session Queue resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.gamelift.GameSessionQueue("test", {
 *     destinations: [
 *         aws_gamelift_fleet.us_west_2_fleet.arn,
 *         aws_gamelift_fleet.eu_central_1_fleet.arn,
 *     ],
 *     playerLatencyPolicies: [
 *         {
 *             maximumIndividualPlayerLatencyMilliseconds: 100,
 *             policyDurationSeconds: 5,
 *         },
 *         {
 *             maximumIndividualPlayerLatencyMilliseconds: 200,
 *         },
 *     ],
 *     timeoutInSeconds: 60,
 * });
 * ```
 *
 * ## Import
 *
 * Gamelift Game Session Queues can be imported by their `name`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:gamelift/gameSessionQueue:GameSessionQueue example example
 * ```
 */
class GameSessionQueue extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["destinations"] = state ? state.destinations : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["playerLatencyPolicies"] = state ? state.playerLatencyPolicies : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["timeoutInSeconds"] = state ? state.timeoutInSeconds : undefined;
        }
        else {
            const args = argsOrState;
            inputs["destinations"] = args ? args.destinations : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["playerLatencyPolicies"] = args ? args.playerLatencyPolicies : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["timeoutInSeconds"] = args ? args.timeoutInSeconds : undefined;
            inputs["arn"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(GameSessionQueue.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing GameSessionQueue resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new GameSessionQueue(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of GameSessionQueue.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === GameSessionQueue.__pulumiType;
    }
}
exports.GameSessionQueue = GameSessionQueue;
/** @internal */
GameSessionQueue.__pulumiType = 'aws:gamelift/gameSessionQueue:GameSessionQueue';
