// @pulumi/aws/route53/resolverRule.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Route53 Resolver rule.
 *
 * ## Example Usage
 * ### System rule
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const sys = new aws.route53.ResolverRule("sys", {
 *     domainName: "subdomain.example.com",
 *     ruleType: "SYSTEM",
 * });
 * ```
 * ### Forward rule
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const fwd = new aws.route53.ResolverRule("fwd", {
 *     domainName: "example.com",
 *     ruleType: "FORWARD",
 *     resolverEndpointId: aws_route53_resolver_endpoint.foo.id,
 *     targetIps: [{
 *         ip: "123.45.67.89",
 *     }],
 *     tags: {
 *         Environment: "Prod",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Route53 Resolver rules can be imported using the `id`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:route53/resolverRule:ResolverRule sys rslvr-rr-0123456789abcdef0
 * ```
 */
class ResolverRule extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["domainName"] = state ? state.domainName : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["ownerId"] = state ? state.ownerId : undefined;
            inputs["resolverEndpointId"] = state ? state.resolverEndpointId : undefined;
            inputs["ruleType"] = state ? state.ruleType : undefined;
            inputs["shareStatus"] = state ? state.shareStatus : undefined;
            inputs["tags"] = state ? state.tags : undefined;
            inputs["targetIps"] = state ? state.targetIps : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.domainName === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'domainName'");
            }
            if ((!args || args.ruleType === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'ruleType'");
            }
            inputs["domainName"] = args ? args.domainName : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["resolverEndpointId"] = args ? args.resolverEndpointId : undefined;
            inputs["ruleType"] = args ? args.ruleType : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["targetIps"] = args ? args.targetIps : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["ownerId"] = undefined /*out*/;
            inputs["shareStatus"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(ResolverRule.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ResolverRule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ResolverRule(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResolverRule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResolverRule.__pulumiType;
    }
}
exports.ResolverRule = ResolverRule;
/** @internal */
ResolverRule.__pulumiType = 'aws:route53/resolverRule:ResolverRule';
