// @pulumi/aws/route53/resolverEndpoint.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Route 53 Resolver endpoint resource.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const foo = new aws.route53.ResolverEndpoint("foo", {
 *     direction: "INBOUND",
 *     securityGroupIds: [
 *         aws_security_group.sg1.id,
 *         aws_security_group.sg2.id,
 *     ],
 *     ipAddresses: [
 *         {
 *             subnetId: aws_subnet.sn1.id,
 *         },
 *         {
 *             subnetId: aws_subnet.sn2.id,
 *             ip: "10.0.64.4",
 *         },
 *     ],
 *     tags: {
 *         Environment: "Prod",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 *  Route 53 Resolver endpoints can be imported using the Route 53 Resolver endpoint ID, e.g.
 *
 * ```sh
 *  $ pulumi import aws:route53/resolverEndpoint:ResolverEndpoint foo rslvr-in-abcdef01234567890
 * ```
 */
class ResolverEndpoint extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["direction"] = state ? state.direction : undefined;
            inputs["hostVpcId"] = state ? state.hostVpcId : undefined;
            inputs["ipAddresses"] = state ? state.ipAddresses : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["securityGroupIds"] = state ? state.securityGroupIds : undefined;
            inputs["tags"] = state ? state.tags : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.direction === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'direction'");
            }
            if ((!args || args.ipAddresses === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'ipAddresses'");
            }
            if ((!args || args.securityGroupIds === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'securityGroupIds'");
            }
            inputs["direction"] = args ? args.direction : undefined;
            inputs["ipAddresses"] = args ? args.ipAddresses : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["securityGroupIds"] = args ? args.securityGroupIds : undefined;
            inputs["tags"] = args ? args.tags : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["hostVpcId"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(ResolverEndpoint.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ResolverEndpoint resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ResolverEndpoint(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResolverEndpoint.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResolverEndpoint.__pulumiType;
    }
}
exports.ResolverEndpoint = ResolverEndpoint;
/** @internal */
ResolverEndpoint.__pulumiType = 'aws:route53/resolverEndpoint:ResolverEndpoint';
